### ISSUE TITLE
io_uring_wait_cqe hangs after read_fixed on /dev/urandom

### ISSUE THREAD

**Issue by JoelJaeschke** [axboe/liburing#549]
I hope this is not a duplicate, but the closest thing I have found is #184. 

Whenever I do a fixed read with registered buffers on something like /dev/urandom or /dev/zero, waiting on the CQE hangs. The program is then left in a Dl state while taking up 100% of a CPU core. I have attached a sample program with which I can reliably reproduce this bug along with a strace.

I have tried setting the IOSQE_ASYNC flag as in #184, but to no avail. When reading from a normal file instead of /dev/urandom, everything works as expected. Is this the expected behavior and am I just missing something?

```
#include <stdio.h>
#include <string.h>
#include <fcntl.h>
#include <stdlib.h>
#include "liburing.h"

#define BUF_SIZE    4096
#define BUFFERS     1
#define IN_FD       "/dev/urandom"
#define OUT_FD      "/dev/zero"

int fixed_buffers(struct io_uring *ring) {
    struct iovec iov[BUFFERS];
    struct io_uring_sqe *sqe;
    struct io_uring_cqe *cqe;

    int fd_in = open(IN_FD, O_RDONLY, 0644);
    if (fd_in < 0 ) {
        perror("open");
        return 1;
    }
    
    int fd_out = open(OUT_FD, O_RDWR, 0644);
    if (fd_out < 0 ) {
        perror("open");
        return 1;
    }

    for (int i = 0; i < BUFFERS; i++) {
        iov[i].iov_base = malloc(BUF_SIZE);
        iov[i].iov_len = BUF_SIZE;
        memset(iov[i].iov_base, 0, BUF_SIZE);
    }

    int ret = io_uring_register_buffers(ring, iov, BUFFERS);
    if(ret) {
        fprintf(stderr, "Error registering buffers: %s", strerror(-ret));
        return 1;
    }

    sqe = io_uring_get_sqe(ring);
    if (!sqe) {
        fprintf(stderr, "Could not get SQE.\n");
        return 1;
    }

    io_uring_prep_read_fixed(sqe, fd_in, iov[0].iov_base, BUF_SIZE, 0, 0);
    io_uring_submit(ring);

    ret = io_uring_wait_cqe(ring, &cqe);
    if (ret < 0) {
        fprintf(stderr, "Error waiting for completion: %s\n",
	        strerror(-ret));
        return 1;
    }
    
    /* Now that we have the CQE, let's process the data */
    if (cqe->res < 0) { 
        fprintf(stderr, "Error in async operation: %s\n", strerror(-cqe->res));
    }
    printf("Result of the operation: %d\n", cqe->res);
    io_uring_cqe_seen(ring, cqe);
    
    sqe = io_uring_get_sqe(ring);
    if (!sqe) {
        fprintf(stderr, "Could not get SQE.\n");
        return 1;
    }
    
    io_uring_prep_write_fixed(sqe, fd_out, iov[0].iov_base, BUF_SIZE, 0, 0);
    io_uring_submit(ring);

    ret = io_uring_wait_cqe(ring, &cqe);
    if (ret < 0) {
        fprintf(stderr, "Error waiting for completion: %s\n",
                strerror(-ret));
        return 1;
    }
    /* Now that we have the CQE, let's process the data */
    if (cqe->res < 0) {
        fprintf(stderr, "Error in async operation: %s\n", strerror(-cqe->res));
    }
    printf("Result of the operation: %d\n", cqe->res);
    io_uring_cqe_seen(ring, cqe);
    
    return 0;
}

int main() {
    struct io_uring ring;

    int ret = io_uring_queue_init(8, &ring, 0);
    if (ret) {
        fprintf(stderr, "Unable to setup io_uring: %s\n", strerror(-ret));
        return 1;
    }
    fixed_buffers(&ring);
    io_uring_queue_exit(&ring);
    return 0;
}
```

Strace:
```
execve("./fix_buf", ["./fix_buf"], 0x7ffd3abebbb0 /* 57 vars */) = 0
brk(NULL)                               = 0x562ffa344000
arch_prctl(0x3001 /* ARCH_??? */, 0x7ffd94a84d10) = -1 EINVAL (Invalid argument)
access("/etc/ld.so.preload", R_OK)      = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, "/etc/ld.so.cache", O_RDONLY|O_CLOEXEC) = 3
newfstatat(3, "", {st_mode=S_IFREG|0644, st_size=47019, ...}, AT_EMPTY_PATH) = 0
mmap(NULL, 47019, PROT_READ, MAP_PRIVATE, 3, 0) = 0x7f99f1555000
close(3)                                = 0
openat(AT_FDCWD, "/lib/liburing.so", O_RDONLY|O_CLOEXEC) = 3
read(3, "\177ELF\2\1\1\0\0\0\0\0\0\0\0\0\3\0>\0\1\0\0\0 \"\0\0\0\0\0\0"..., 832) = 832
newfstatat(3, "", {st_mode=S_IFREG|0755, st_size=79864, ...}, AT_EMPTY_PATH) = 0
mmap(NULL, 8192, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_ANONYMOUS, -1, 0) = 0x7f99f1553000
mmap(NULL, 24736, PROT_READ, MAP_PRIVATE|MAP_DENYWRITE, 3, 0) = 0x7f99f154c000
mmap(0x7f99f154e000, 8192, PROT_READ|PROT_EXEC, MAP_PRIVATE|MAP_FIXED|MAP_DENYWRITE, 3, 0x2000) = 0x7f99f154e000
mmap(0x7f99f1550000, 4096, PROT_READ, MAP_PRIVATE|MAP_FIXED|MAP_DENYWRITE, 3, 0x4000) = 0x7f99f1550000
mmap(0x7f99f1551000, 8192, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_FIXED|MAP_DENYWRITE, 3, 0x4000) = 0x7f99f1551000
close(3)                                = 0
openat(AT_FDCWD, "/lib/x86_64-linux-gnu/libc.so.6", O_RDONLY|O_CLOEXEC) = 3
read(3, "\177ELF\2\1\1\3\0\0\0\0\0\0\0\0\3\0>\0\1\0\0\0000\242\2\0\0\0\0\0"..., 832) = 832
pread64(3, "\6\0\0\0\4\0\0\0@\0\0\0\0\0\0\0@\0\0\0\0\0\0\0@\0\0\0\0\0\0\0"..., 784, 64) = 784
pread64(3, "\4\0\0\0 \0\0\0\5\0\0\0GNU\0\2\0\0\300\4\0\0\0\3\0\0\0\0\0\0\0"..., 48, 848) = 48
pread64(3, "\4\0\0\0\24\0\0\0\3\0\0\0GNU\0\360\374)\26\\\276`\210\300\341\255\360;\0H\373"..., 68, 896) = 68
newfstatat(3, "", {st_mode=S_IFREG|0644, st_size=2216272, ...}, AT_EMPTY_PATH) = 0
pread64(3, "\6\0\0\0\4\0\0\0@\0\0\0\0\0\0\0@\0\0\0\0\0\0\0@\0\0\0\0\0\0\0"..., 784, 64) = 784
mmap(NULL, 2260560, PROT_READ, MAP_PRIVATE|MAP_DENYWRITE, 3, 0) = 0x7f99f1324000
mprotect(0x7f99f134c000, 2019328, PROT_NONE) = 0
mmap(0x7f99f134c000, 1654784, PROT_READ|PROT_EXEC, MAP_PRIVATE|MAP_FIXED|MAP_DENYWRITE, 3, 0x28000) = 0x7f99f134c000
mmap(0x7f99f14e0000, 360448, PROT_READ, MAP_PRIVATE|MAP_FIXED|MAP_DENYWRITE, 3, 0x1bc000) = 0x7f99f14e0000
mmap(0x7f99f1539000, 24576, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_FIXED|MAP_DENYWRITE, 3, 0x214000) = 0x7f99f1539000
mmap(0x7f99f153f000, 52816, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_FIXED|MAP_ANONYMOUS, -1, 0) = 0x7f99f153f000
close(3)                                = 0
mmap(NULL, 12288, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_ANONYMOUS, -1, 0) = 0x7f99f1321000
arch_prctl(ARCH_SET_FS, 0x7f99f1321740) = 0
set_tid_address(0x7f99f1321a10)         = 1564
set_robust_list(0x7f99f1321a20, 24)     = 0
mprotect(0x7f99f1539000, 16384, PROT_READ) = 0
mprotect(0x7f99f1551000, 4096, PROT_READ) = 0
mprotect(0x562ff9d53000, 4096, PROT_READ) = 0
mprotect(0x7f99f1594000, 8192, PROT_READ) = 0
prlimit64(0, RLIMIT_STACK, NULL, {rlim_cur=8192*1024, rlim_max=RLIM64_INFINITY}) = 0
munmap(0x7f99f1555000, 47019)           = 0
io_uring_setup(8, {flags=0, sq_thread_cpu=0, sq_thread_idle=0, sq_entries=8, cq_entries=16, features=IORING_FEAT_SINGLE_MMAP|IORING_FEAT_NODROP|IORING_FEAT_SUBMIT_STABLE|IORING_FEAT_RW_CUR_POS|IORING_FEAT_CUR_PERSONALITY|IORING_FEAT_FAST_POLL|IORING_FEAT_POLL_32BITS|IORING_FEAT_SQPOLL_NONFIXED|IORING_FEAT_EXT_ARG|IORING_FEAT_NATIVE_WORKERS|IORING_FEAT_RSRC_TAGS, sq_off={head=0, tail=64, ring_mask=256, ring_entries=264, flags=276, dropped=272, array=576}, cq_off={head=128, tail=192, ring_mask=260, ring_entries=268, overflow=284, cqes=320, flags=0x118 /* IORING_CQ_??? */}}) = 3
mmap(NULL, 608, PROT_READ|PROT_WRITE, MAP_SHARED|MAP_POPULATE, 3, 0) = 0x7f99f1560000
mmap(NULL, 512, PROT_READ|PROT_WRITE, MAP_SHARED|MAP_POPULATE, 3, 0x10000000) = 0x7f99f155f000
openat(AT_FDCWD, "/dev/urandom", O_RDONLY) = 4
openat(AT_FDCWD, "/dev/zero", O_RDWR)   = 5
getrandom("\xb2\xa0\x40\x02\x2b\x30\xd9\x23", 8, GRND_NONBLOCK) = 8
brk(NULL)                               = 0x562ffa344000
brk(0x562ffa365000)                     = 0x562ffa365000
io_uring_register(3, IORING_REGISTER_BUFFERS, [{iov_base=0x562ffa3442a0, iov_len=4096}], 1) = 0
io_uring_enter(3, 1, 0, 0, NULL, 8)     = 1
io_uring_enter(3, 0, 1, IORING_ENTER_GETEVENTS, NULL, 8^C) = -1 EINTR (Interrupted system call)
```

Platform: occurs on Debian with kernel 5.10.0-8-amd64 with liburing 0.7-3 and a freshly setup Ubuntu VM with kernel 5.13.0-35-generic and liburing 2.2 compiled from eb02ccc

**Comment by axboe** [comment]
Thanks for the report! I'll take a look.

**Comment by axboe** [comment]
Fix here:

https://git.kernel.dk/cgit/linux-block/commit/?h=for-5.18/io_uring&id=7ee7ca69d5f9ac5959072d9cc85cd31706ea72cd

If you would verify your full name and email for me, I can update the Reported-by tag to be correct.

Also would you mind if I adopt your test program as a liburing regression test?

**Comment by JoelJaeschke** [comment]
Damn, you are fast!

Full name is Joel Jaeschke. My email is `joel.jaeschke at gmail dot com`

Yeah sure, happy to help.

**Comment by axboe** [comment]
Great thanks! Updated the commit, and added the liburing test case too.

> Damn, you are fast!

A bug report with a reproducer is a blessing, makes fixing it soooo easy. So thanks for that!

**Comment by axboe** [comment]
Fix will go into 5.18 and be backported to stable next week.

**Comment by JoelJaeschke** [comment]
Awesome, thank you for your work :)
